src/core/Akka.Streams/Implementation/ChannelSources.cs

```diff
@@ -0,0 +1,155 @@
+﻿// //-----------------------------------------------------------------------
+// // <copyright file="ChannelReaderSource.cs" company="Akka.NET Project">
+// //     Copyright (C) 2009-2022 Lightbend Inc. <http://www.lightbend.com>
+// //     Copyright (C) 2013-2022 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Akka.Streams.Stage;
+
+namespace Akka.Streams.Implementation
+{
+    sealed class ChannelSourceLogic<T> : OutGraphStageLogic
+    {
+        private readonly Outlet<T> _outlet;
+        private readonly ChannelReader<T> _reader;
+        private readonly Action<bool> _onValueRead;
+        private readonly Action<Exception> _onValueReadFailure;
+        private readonly Action<Exception> _onReaderComplete;
+        private readonly Action<Task<bool>> _onReadReady;
+
+        public ChannelSourceLogic(SourceShape<T> source, Outlet<T> outlet,
+            ChannelReader<T> reader) : base(source)
+        {
+            _outlet = outlet;
+            _reader = reader;
+            _onValueRead = GetAsyncCallback<bool>(OnValueRead);
+            _onValueReadFailure =
+                GetAsyncCallback<Exception>(OnValueReadFailure);
+            _onReaderComplete = GetAsyncCallback<Exception>(OnReaderComplete);
+            _onReadReady = ContinueAsyncRead;
+            _reader.Completion.ContinueWith(t =>
+            {
+                if (t.IsFaulted) _onReaderComplete(t.Exception);
+                else if (t.IsCanceled)
+                    _onReaderComplete(new TaskCanceledException(t));
+                else _onReaderComplete(null);
+            });
+
+            SetHandler(_outlet, this);
+        }
+
+        private void OnReaderComplete(Exception reason)
+        {
+            if (reason is null)
+                CompleteStage();
+            else
+                FailStage(reason);
+        }
+
+        private void OnValueReadFailure(Exception reason) => FailStage(reason);
+
+        private void OnValueRead(bool dataAvailable)
+        {
+            if (dataAvailable && _reader.TryRead(out var element))
+                Push(_outlet, element);
+            else
+                CompleteStage();
+        }
+
+        public override void OnPull()
+        {
+            if (_reader.TryRead(out var element))
+            {
+                Push(_outlet, element);
+            }
+            else
+            {
+                var continuation = _reader.WaitToReadAsync();
+                if (continuation.IsCompletedSuccessfully)
+                {
+                    var dataAvailable = continuation.GetAwaiter().GetResult();
```

Comment:
Any way we can use async callbacks here?

Reply:
I would need a bit more context to answer that?

AFAIK this code path is present to specifically avoid `GetAsyncCallBack` and being put on scheduler.

Reply:
Got it.